core/trino-main/src/main/java/io/trino/execution/UseTask.java

```diff
@@ -64,14 +69,17 @@ public ListenableFuture<Void> execute(
                 .orElseGet(() -> session.getCatalog().orElseThrow(() ->
                         semanticException(MISSING_CATALOG_NAME, statement, "Catalog must be specified when session catalog is not set")));
 
-        if (metadata.getCatalogHandle(session, catalog).isEmpty()) {
+        SecurityContext securityContext = session.toSecurityContext();
+        if (hasCatalogAccess(securityContext, catalog) ||
+                metadata.getCatalogHandle(session, catalog).isEmpty()) {
             throw new TrinoException(NOT_FOUND, "Catalog does not exist: " + catalog);
         }
 
         String schema = statement.getSchema().getValue().toLowerCase(ENGLISH);
 
         CatalogSchemaName name = new CatalogSchemaName(catalog, schema);
-        if (!metadata.schemaExists(session, name)) {
+        if (hasSchemaAccess(securityContext, catalog, schema) ||
+                !metadata.schemaExists(session, name)) {
             throw new TrinoException(NOT_FOUND, "Schema does not exist: " + name);
```

Comment:
I think the error message has to be different and so the error code should be different between not existing case or access denied case.

Reply:
IMO error message needs to be unambiguous. With different error messages there will be no difference what we have today.

Reply:
That is the way Trino works today. For example see `io.trino.sql.analyzer.StatementAnalyzer.Visitor#visitUpdate` if you use not existing table you will get different error than when using a table without an access.

Reply:
Ok, I check if catalog/schema exists first, then I check if user has an access to those